policy-controller/k8s/index/src/http_route.rs

```diff
@@ -0,0 +1,136 @@
+use k8s_gateway_api::ParentReference;
+use linkerd_policy_controller_core::http_route::{
+    HeaderMatch, Hostname, HttpMethod, HttpRoute, HttpRouteMatch, PathMatch, QueryParamMatch, Value,
+};
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct RouteBinding {
+    pub route: HttpRoute,
+    pub parent_refs: Vec<ParentReference>,
+}
+
+impl RouteBinding {
+    pub fn from_resource(route: k8s_gateway_api::HttpRoute) -> Self {
+        let hostnames = route
+            .spec
+            .hostnames
+            .iter()
+            .flatten()
+            .map(|hostname| {
+                if hostname.starts_with("*.") {
+                    let mut reverse_labels = hostname
+                        .split('.')
+                        .skip(1)
+                        .map(|label| label.to_owned())
+                        .collect::<Vec<String>>();
+                    reverse_labels.reverse();
+                    Hostname::Suffix { reverse_labels }
+                } else {
+                    Hostname::Exact(hostname.to_owned())
+                }
+            })
+            .collect();
+
+        let matches = route
+            .spec
+            .rules
+            .iter()
+            .flatten()
+            .filter_map(|rule| rule.matches.as_ref())
+            .flatten()
+            .map(|route_match| {
```

Comment:
I'm still working through some other suggestions, but this jumps out to me. I think this is wrong: when the proxy evaluates a rule, it checks that _all_ matches in a rule match before applying its filters. By flattening the matches from all rules, we're changing the semantics--the proxy will only apply the route if all matches from all rules match.

Also, we're dropping filters on the floor here, and we'll need to preserve them.